File: impls/js2/step8_macros.js
import { Env, makeClosureEnv } from './env.js'
import { read_str } from './reader.js'
import { pr_str } from './printer.js'
import {
  repl_env,
  list,
  isList,
  isSymbol,
  vector,
  hash_map,
  symbol,
  isVector,
  isHashMap,
  isClosure,
  cloneFunction,
} from './core.js'

const READ = str => read_str(str)

const eval_ast = (ast, env) => {
  if (ast === null) return null
  const { type, value } = ast
  switch (type) {
    case 'symbol':
      return env.get(value)
    case 'list':
      return list(...ast.value.map(x => EVAL(x, env)))
    case 'vector':
      return vector(...ast.value.map(x => EVAL(x, env)))
    case 'hash-map':
      return hash_map(
        ...[].concat(
          ...[...ast.value.entries()].map(([k, v]) => [k, EVAL(v, env)]),
        ),
      )
    default:
      return ast
  }
}

const quasiquote = ast => {
  if (ast === undefined) return null
  if (isList(ast)) {
    const elements = ast.value
    if (elements.length === 0) return ast
    if (isSymbol(elements[0]) === 'unquote' && elements.length >= 2)
      return elements[1]
    const [elt, ...elts] = elements
    if (isList(elt) && isSymbol(elt.value[0]) === 'splice-unquote') {
      return list(symbol('concat'), elt.value[1], quasiquote(list(...elts)))
    } else {
      return list(symbol('cons'), quasiquote(elt), quasiquote(list(...elts)))
    }
  } else if (isVector(ast)) {
    if (isSymbol(ast.value[0]) === 'unquote') {
      return list(
        symbol('vec'),
        list(
          symbol('cons'),
          list(symbol('quote'), symbol('unquote')),
          quasiquote(list(...ast.value.slice(1))),
        ),
      )
    } else {
      return list(symbol('vec'), quasiquote(list(...ast.value)))
    }
  }
  if (isHashMap(ast) || isSymbol(ast)) return list(symbol('quote'), ast)
  return ast
}

const isMacroCall = (ast, env) => {
  try {
    if (isList(ast)) {
      const [a] = ast.value
      return isSymbol(a) && env.get(a.value)?.isMacro
    }
  } catch (e) {
    return false
  }
  return false
}

const macroExpand = (ast, env) => {
  while (isMacroCall(ast, env)) {
    const [a, ...args] = ast.value
    const macro = env.get(a.value)
    ast = macro(...args)
  }
  return ast
}

const EVAL = (ast, env) => {
  while (true) {
    if (ast === null) return null
    if (!isList(ast)) return eval_ast(ast, env)
    if (ast.value.length === 0) return ast
    ast = macroExpand(ast, env)
    if (!isList(ast)) continue
    const [efirst, ...rest] = ast.value
    if (isSymbol(efirst)) {
      const efirstValue = efirst.value
      switch (efirstValue) {
        case 'def!':
        case 'defmacro!': {
          const [symbol, value] = rest
          let ev = EVAL(value, env)
          if (efirstValue === 'defmacro!' && typeof ev === 'function') {
            ev = cloneFunction(ev)
            ev.isMacro = true
          }
          env.set(symbol.value, ev)
          return ev
        }
        case 'let*': {
          const [bindings, body] = rest
          const newEnv = new Env(env)
          for (let i = 0; i < bindings.value.length; i += 2) {
            newEnv.set(
              bindings.value[i].value,
              EVAL(bindings.value[i + 1], newEnv),
            )
          }
          ast = body
          env = newEnv
          continue
        }
        case 'do': {
          for (const x of rest.slice(0, -1)) EVAL(x, env)
          ast = rest.at(-1)
          continue
        }
        case 'if': {
          const [cond, t, f] =
            rest.length < 3 ? rest.concat(null, null, null) : rest
          const econd = EVAL(cond, env)
          ast = econd === false || econd === null ? f : t
          continue
        }
        case 'fn*': {
          const [params, body] = rest
          const closureCtor = makeClosureEnv(params, env)
          const fn = (...args) => EVAL(body, closureCtor(args))
          fn.type = 'closure'
          fn.ast = body
          fn.closureCtor = closureCtor
          return fn
        }
        case 'quote': {
          return rest[0]
        }
        case 'quasiquote': {
          ast = quasiquote(rest[0])
          continue
        }
        case 'quasiquoteexpand': {
          return quasiquote(rest[0])
        }
        case 'macroexpand': {
          return macroExpand(rest[0], env)
        }
      }
    }
    const f = EVAL(efirst, env)
    const args = rest.map(arg => EVAL(arg, env))
    if (f.type !== 'closure') return f(...args)
    ast = f.ast
    env = f.closureCtor(args)
  }
}

const PRINT = str => pr_str(str, true)

const rep = (str, env) => PRINT(EVAL(READ(str), env))

const env = new Env()

for (const [key, value] of Object.entries(repl_env)) env.set(key, value)

rep(`(def! not (fn* (a) (if a false true)))`, env)

env.set('eval', ast => EVAL(ast, env))

rep(
  `(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \"\nnil)\")))))`,
  env,
)

rep(
  '(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list \'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw "odd number of forms to cond")) (cons \'cond (rest (rest xs)))))))',
  env,
)

const commandLineArgs = process.argv.slice(2)
const [first, ...rest] = commandLineArgs
env.set('*ARGV*', list(...rest))

import * as readline from 'node:readline'
import { stdin as input, nextTick, stdout as output } from 'node:process'

if (commandLineArgs.length > 0) {
  rep(`(load-file "${first}")`, env)
} else {
  const rl = readline.createInterface({ input, output })
  const prompt = () => {
    rl.question(`user> `, line => {
      if (line === '') {
        console.log(`Bye!`)
        rl.close()
        return
      }
      try {
        console.log(rep(line, env))
      } catch (e) {
        console.log(e.message)
      }
      nextTick(prompt)
    })
  }
  prompt()
}

File: impls/clojure.2/src/mal/step8_macros.clj
(ns mal.step8-macros
  (:refer-clojure :exclude [macroexpand])
  (:require
   [mal.printer]
   [mal.reader]
   [mal.env]
   [mal.core]))

(defn READ [s]
  (mal.reader/read-form s))

(declare EVAL)

(defn eval-ast [ast env]
  (cond
    (symbol? ast)
    (mal.env/get-symbol ast env)

    (list? ast)
    (map #(EVAL % env) ast)

    (vector? ast)
    (mapv #(EVAL % env) ast)

    (map? ast)
    (into {} (for [[k v] ast] [k (EVAL v env)]))

    :else
    ast))

(defn mal-apply [f args]
  (if (fn? f)
    (apply f args)
    (let [{:keys [ast params env]} f
          new-env (mal.env/make-env env params args)]
      (EVAL ast new-env))))

(defn quasiquote [ast]
  (cond
    (list? ast)
    (cond
      (empty? ast)
      ast

      (and (= (first ast) 'unquote) (not (nil? (second ast))))
      (second ast)

      :else
      (let [elt (first ast)
            elts (rest ast) ; use rest instead of destructuring to get () not nil on empty
            ]
        (if (and (list? elt) (= 'splice-unquote (first elt)))
          (list 'concat (second elt) (quasiquote elts))
          (list 'cons (quasiquote elt) (quasiquote elts)))))

    (or (map? ast) (symbol? ast))
    (list 'quote ast)

    (vector? ast)
    (if (= (first ast) 'unquote)
      (list 'vec (list 'cons '(quote unquote) (quasiquote (apply list (rest ast)))))
      (list 'vec (quasiquote (apply list ast))))

    :else
    ast))

(defn macro-call? [ast env]
  (and (list? ast)
       (symbol? (first ast))
       (:is-macro (mal.env/try-get-symbol (first ast) env))))

(defn macroexpand [call-ast env]
  (if (macro-call? call-ast env)
    (let [[first-form & args] call-ast
          {:keys [ast params env]} (mal.env/get-symbol first-form env)
          new-env (mal.env/make-env env params args)]
      (EVAL ast new-env))
    call-ast))

(defn EVAL [ast env]
  (if (list? ast)
    (if (empty? ast)
      ast
      (let [ast (macroexpand ast env)]
        (if-not (list? ast)
          (eval-ast ast env)
          (let [[first-form & forms] ast]
            (cond
              (= first-form 'def!)
              (let [[name exp] forms
                    evaluated-exp (EVAL exp env)]
                (mal.env/set-symbol name evaluated-exp env)
                evaluated-exp)

              (= first-form 'defmacro!)
              (let [[name exp] forms
                    evaluated-exp (EVAL exp env)
                    macro-exp (assoc evaluated-exp :is-macro true)]
                (mal.env/set-symbol name macro-exp env)
                macro-exp)

              (= first-form 'let*)
              (let [[bindings body] forms
                    new-env (mal.env/make-env env)]
                (doseq [[name exp] (partition 2 bindings)]
                  (mal.env/set-symbol name (EVAL exp new-env) new-env))
                (recur body new-env))

              (= first-form 'do)
              (do
                (doseq [f (butlast forms)] (EVAL f env))
                (recur (last forms) env))

              (= first-form 'if)
              (let [[condition true-form false-form] forms]
                (if (EVAL condition env)
                  (recur true-form env)
                  (recur false-form env)))

              (= first-form 'fn*)
              (let [[params body] forms]
                (mal.printer/map->Closure
                 {:ast body
                  :params params
                  :env env}))

              (= first-form 'quote)
              (first forms)

              (= first-form 'quasiquoteexpand)
              (quasiquote (first forms))

              (= first-form 'quasiquote)
              (recur (quasiquote (first forms)) env)

              (= first-form 'macroexpand)
              (macroexpand (first forms) env)

              :else
              (let [f (EVAL first-form env)
                    args (for [form forms] (EVAL form env))]
                (if (fn? f)
                  (apply f args)
                  (recur (:ast f) (mal.env/make-closure-env f args)))))))))
    (eval-ast ast env)))

(defn PRINT [s]
  (mal.printer/pr-str s))

(defn rep [input env]
  (->
   input
   READ
   (EVAL env)
   PRINT))

(def repl-env
  (mal.env/make-env))

(def mal-ns
  {'eval (fn [ast] (EVAL ast repl-env))
   'swap! (fn [a f & params] (apply swap! a (fn [& ps] (mal-apply f ps)) params))})

(doseq [[s f] (merge mal.core/ns mal-ns)]
  (mal.env/set-symbol s f repl-env))

(def string-defs
  ["(def! not (fn* (a) (if a false true)))"
   "(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \"\nnil)\")))))"
   "(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))"])

(doseq [s string-defs]
  (rep s repl-env))

(mal.env/set-symbol '*ARGV* (apply list (rest *command-line-args*)) repl-env)

(if *command-line-args*
  (rep (str "(load-file \"" (first *command-line-args*) "\")") repl-env)
  (loop []
    (print "user> ")
    (flush)
    (when-let [line (read-line)]
      (try
        (println (rep line repl-env))
        (catch Exception e
          (println (str "error:" (ex-message e)))))
      (recur))))
